WIP : Channel Adapters for Redis Streams

akuma8 opened the pull request:
#3226 
The PR is in WIP because I need information for improvements.

akuma8 committed:
Adding Channel Adapters for Redis Streams

akuma8 commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/inbound/RedisStreamInboundChannelAdapter.java:
```diff
@@ -0,0 +1,272 @@
+
+    private Message<?> convertMessage( Record<Object, Object> message ) {
+        return this.messageConverter.toMessage( Record.of( message ), null );
+    }
```
@artembilan How to improve this converter? 
Currently I send back a `Record` to the output channel, it would be ideal to extract the paylod and headers and construct a real `Message`  before publishing it.

akuma8 commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/outbound/RedisStreamOutboundChannelAdapter.java:
```diff
@@ -0,0 +1,58 @@
+        ObjectRecord<String, Message<?>> record = StreamRecords
+                .<String, Message<?>> objectBacked( message )
+                .withStreamKey( streamName );
+        this.streamOperations.add( record );
```
@artembilan How can I improve this code?
Currently I don't make any difference about incoming data type. I also publish all the message as is because I want to keep its headers.
Also I have some troubles about how well converting and reading data. 
Say I have that class:
```
public class SimpleUser implements Serializable {
    private final String nom;
    private final String prenom;
    private final String username;
}
```
When publishing a message with a paylod of type `SimpleUser ` I am unable to get it from the inbound part (see my comment above). In the inbound side, I would like to extract it before sending it to the output channel.

artembilan reviewed (changes requested):
I hope you find my superficial review sufficient and we are good to continue a work here.

Thank you for your effort!

artembilan commented:
@akuma8 ,

please, don't squash commits when you push them to the PR: it is going to be much easier for us to review when change history is present.
Also it would be great to comment the changes in commit: kinda what has been done and the reason.
It also helps us to review faster.

See more info about good commits in Chris Beams' article: https://chris.beams.io/posts/git-commit/

Thanks for understanding.

P.S. for now it is OK because PR is small yet, but it is going to grow I believe.

akuma8 commented:
@artembilan Thank you for the remark, I will be careful in the future. It's a personal habit.
I let you a comment because I am stuck, please let me know when you will read it.

artembilan commented:
>  please let me know when you will read it.

I'm in review process. 
Stay tuned. BTW not reason to be in hurry: we are OK to release it whenever it is ready.

artembilan commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/inbound/RedisStreamInboundChannelAdapter.java:
```diff
@@ -0,0 +1,155 @@
+    @Override
+    protected void doStop() {
+        super.doStop();
+        this.container.stop();
```
I think `container` has to be stopped first.

artembilan commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/inbound/RedisStreamInboundChannelAdapter.java:
```diff
@@ -0,0 +1,155 @@
+        this.createGroupeIfNotExist = createGroupeIfNotExist;
+    }
+
+    public void setReadOffset( ReadOffset readOffset ) {
```
Setters/getters have to go after ctors.

artembilan commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/inbound/RedisStreamInboundChannelAdapter.java:
```diff
@@ -0,0 +1,155 @@
+    public void setReadOffset( ReadOffset readOffset ) {
+        this.readOffset = readOffset;
+    }
+}
```
Every class member has to be wrapped with blank lines.

artembilan commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/inbound/RedisStreamInboundChannelAdapter.java:
```diff
@@ -0,0 +1,155 @@
+package org.springframework.integration.redis.inbound;
```
Copyright has to be present in source files. In classes to be precise.

artembilan commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/outbound/RedisStreamMessageHandler.java:
```diff
@@ -0,0 +1,90 @@
+
+	public RedisStreamMessageHandler( String streamKey, RedisConnectionFactory connectionFactory ) {
+		Assert.notNull( streamKey, "'streamName' must not be null" );
+		Assert.notNull( connectionFactory, "'connectionFactory' must not be null" );
```
We don't need this because `template.afterPropertiesSet();` ensures such an assertion.

artembilan commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/outbound/RedisStreamMessageHandler.java:
```diff
@@ -0,0 +1,90 @@
+	}
+
+	public void setStreamKeyExpression( Expression streamKeyExpression ) {
+		this.streamKeyExpression = streamKeyExpression;
```
This cannot be `null`. So, an appropriate assert.

Also I think this is really an important option, so has to be in the ctor.
I'm OK with an existing ctor for the `String streamKey`, but then we need to have a new one based on the ` Expression streamKeyExpression`.
Therefore `final` for expression property and not setters at all.

artembilan commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/outbound/RedisStreamMessageHandler.java:
```diff
@@ -0,0 +1,90 @@
+
+	private volatile Expression streamKeyExpression;
+
+	private volatile boolean extractPayload = true;
```
As I said before: no `volatile` for properties which are not changed at run time.

artembilan commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/outbound/RedisStreamMessageHandler.java:
```diff
@@ -0,0 +1,90 @@
+	@Override
+	protected void handleMessageInternal( Message<?> message ) {
+
+		String streamKey = this.streamKeyExpression.getExpressionString();
```
That's something awkward.
You probably need to make yourself familiar with an `Expression.getValue()`.
See other `MessageHandler` implementations in this Redis module to determine how we evaluate similar SpELs.

artembilan reviewed (changes requested):
You know I have just discovered for myself a `ReactiveStreamOperations` in Spring Data Redis.
I my experience it is better to develop Reactive component and let it to be used in the blocking scenarios.

Can we revise your solution in favor of `ReactiveStreamOperations`?
See `ReactiveMongoDbStoringMessageHandler` for example.

The `RedisStreamInboundChannelAdapter` could also be based on the `ReactiveStreamOperations.read()`. The returned `Flux` could be damped into a `ReactiveStreamsSubscribableChannel` if configured that way or as regular `send()` otherwise.

Let me know WDYT?

akuma8 reviewed (commented):
> Can we revise your solution in favor of `ReactiveStreamOperations`?

Sure, I saw `ReactiveStreamOperations` in the doc but I thought it only be relevant to use it in a full reactive context which is not our case but I will revise my first solution if this one is better.

akuma8 commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/inbound/RedisStreamInboundChannelAdapter.java:
```diff
@@ -0,0 +1,155 @@
+            .pollTimeout( Duration.ZERO )
+            .executor( new SimpleAsyncTaskExecutor() )
+            .batchSize( 1 )
+            .objectMapper( new ObjectHashMapper() )
```
I simply didn't see it, my mistake.

About formatting, I will do it at the end, when we will have something that works as expected.

akuma8 committed:
Inbound:
 - Changed the imperative `StreamMessageListenerContainer` listener in favor of the reactive one `StreamReceiver`
 - Added RedisHeaders values for Redis Stream

Outbound:
 - Added a new constructor taking `Expression` as parameter
 - Changed the imperative `StreamOperations` in favor of `ReactiveStreamOperations`
 - Renaming class from `RedisStreamMessageHandler` to `ReactiveRedisStreamMessageHandler`

Took into account other improvement remarks: copyright, setters position, methods names

artembilan commented:
Hi, @akuma8 !

Thank you for update! I'll review your solution and try to figure problems the next week.
Meanwhile the answer to your question: yes, I want an *outbound* part as reactive as well. You need to implement a `ReactiveMessageHandler` over there. See a `ReactiveMongoDbStoringMessageHandler` for example.

Right now I'm working on some reactive streams functionality for the `MessageProducerSupport`. I somehow believe your inbound channel adapter would benefit from that.
Let's see the next week how it is going!

artembilan commented:
I have just raised a PR https://github.com/spring-projects/spring-integration/pull/3240 about reactive Message Producer.
I have over there an implementation for MongoDb change stream.
I think that one should be a sample how this Redis Streams has to be implemented on the inbound side.
Let's hope it will be merged next week, so you would have some foundation to rely on!

artembilan reviewed (changes requested):
@akuma8 ,

tell, me, please, how that would be hard for you to divide this PR into two.
One for already working functionality on the outbound side.
And then we can come back to be inbound one when PR #3240 is merged.
This way we won't have everything blocked.

Thank for understanding.

akuma8 commented:
Hi @artembilan
Since Monday I am sick, it’s not the Covid-19 😁.
It’s a good idea to split this PR in 2, I’ll do that next week, I hope getting better.

artembilan commented:
Stay safe, @akuma8 , and get well soon!

No need to be in a hurry: looks like no pressure for the feature yet. So, we are good to bring it whenever it is ready.

When you come back from your sick time off, please, divide this into two PRs - inbound & outbound.
Please, make sure you contribute some tests for this new channel adapters.

At a glance `MessageHandler` part is OK - should be merged shortly after an appropriate PR.

I haven't looked into an inbound part because wanted to be sure that `MessageProducerSupport.subscribeToPublisher(Publisher<Message<?>>)` approach is correct.

Now it is merged to master and you can start looking into the `MongoDbChangeStreamMessageProducer` to make your inbound part similar way.

Fingers crossed for your health!

artembilan reviewed (changes requested):
Hi @akuma8 !

Any chances to see an update from you on the matter?

What we have so far and what we still need to do to have that Redis Stream Inbound Channel Adapter?

Thanks

akuma8 commented:
Hi @artembilan,
I am a bit busy, I started looking at the inbound part, I hope opening a first PR next week. 
I also opened that issue about [reading List from a Redis Stream](https://jira.spring.io/browse/DATAREDIS-1175)
I'll let you know. 
Thanks

artembilan commented:
Hi @akuma8 !

No problem with your load!

Just let me know that you can make it before September or should I pull this task from you and fix myself!
Great to see the fix in Spring Data Redis. I think it's time reinstate your test for list. Nothing more we need to do on the matter.
See a related PR in Spring Data: they have some tests to investigate.

akuma8 commented:
@artembilan,
> ... you can make it before September...

Yes, I think so. If we keep doing it like we did for the outbound side I can make it before September.
I am also looking for a way to test the fix from Spring Data Redis, if you can include the related version, please let me know.

artembilan commented:
>  if you can include the related version, please let me know.

Just did: https://github.com/spring-projects/spring-integration/commit/ebec500abd85d10b9eb584a8794f1abba4acf0f6

Upgraded to Spring Data `2020.0.0-SNAPSHOT`

akuma8 commented:
Hi @artembilan,
I have some questions about implementation choices.
Redis Stream Api provides 2 ways to read data, we can read as a simple client with the `XREAD` command or as a client belonging to a Consumer Group: `XREADGROUP`. Should we provide one inbound implementation including the 2 options or devide them in 2 different implementations? 
Thanks a lot

artembilan commented:
Yes, if that is possible, I'd prefer to have a one inbound component with a `consumerGroup` property (optional).
And yes, it would be better to have it as reactive. I think the `MongoDbChangeStreamMessageProducer` is a good sample to borrow ideas.

Thanks

artembilan commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/inbound/RedisStreamInboundChannelAdapter.java:
```diff
@@ -0,0 +1,171 @@
+			Assert.hasText( consumerName, "'consumerName' must be set" );
+			Consumer consumer = Consumer.from( this.consumerGroupName, this.consumerName );
+			this.receiver
+					.receiveAutoAck( consumer, offset )
```
Looks like we also need to think about manual acknowledgement:
```
 * Every record must be acknowledged using
	 * {@link org.springframework.data.redis.connection.ReactiveStreamCommands#xAck(ByteBuffer, String, String...)}
```

See `receive()` API for the `StreamReceiver`.
This has to be populated as a `IntegrationMessageHeaderAccessor.ACKNOWLEDGMENT_CALLBACK`.
And also see `SimpleAcknowledgment`  abstraction to wrap the call to that `xAck`. for the particular record returned.

Saying that I'd expect an `autoAck(true|false)` option on this channel adapter.

artembilan reviewed (changes requested):
This is good progress, but definitely there is still some work to do: tests, docs, JavaDocs etc.

Thanks

akuma8 commented:
Hi @artembilan,
Sorry but I forgot to tell you to not review this, I just rebased master branch. 
I planned to delete this PR and opening a new one.

artembilan commented:
OK. So, I'm closing this one to avoid confusion.

Thanks for confirmation!
